### PR TITLE
Fix issue when completing the first command in a script with an empty array expression

### PR DESCRIPTION
**Pull request by MartinGC94**
# PR Summary
Fixes tab completion in scenarios like: `get-chi<Tab> @()`

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Comment by iSazonov** [comment]
> Fixes tab completion in scenarios like: `get-chi<Tab> @()`

This works for me without the change.

**Comment by MartinGC94** [comment]
> > Fixes tab completion in scenarios like: `get-chi<Tab> @()`
> 
> This works for me without the change.

Which version are you on? I believe this issue was added with my index expression completion changes and I think they only got added in the most recent 7.3.0 preview 8 release.  
This issue occurs if you are tab completing a command that starts in the very first position of a script that also contains an empty array expression. The empty array expression creates an empty script extent which has the same starting position as a command in the first position would have. If there's any whitespace before the command then it works properly.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs [hunk 1]
```diff
@@ -2036,7 +2036,9 @@ private static List<CompletionResult> GetResultForIdentifier(CompletionContext c
                 // Handle completion for a path with variable, such as: $PSHOME\ty<tab>
                 if (completionContext.RelatedAsts.Count > 0 && completionContext.RelatedAsts[0] is ScriptBlockAst)
                 {
-                    Ast cursorAst = completionContext.RelatedAsts[0].FindAll(ast => ast.Extent.EndOffset <= tokenAtCursor.Extent.StartOffset, true).LastOrDefault();
+                    Ast cursorAst = completionContext.RelatedAsts[0].FindAll(
+                        ast => ast.Extent.EndOffset <= tokenAtCursor.Extent.StartOffset
+                        && ast.Extent is not EmptyScriptExtent, true).LastOrDefault();
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs:2041]:
```suggestion
                        ast => ast.Extent.EndOffset <= tokenAtCursor.Extent.StartOffset
                            && ast.Extent is not EmptyScriptExtent,
                        searchNestedScriptBlocks: true).LastOrDefault();
```

File: test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1 [hunk 2]
```diff
@@ -818,6 +818,11 @@ Verb-Noun -Param1 Hello ^
         $res.CompletionMatches[0].CompletionText | Should -Be "Break"
     }
 
+    it 'Should complete command with an empty arrayexpression element' {
+        $res = TabExpansion2 -inputScript 'Get-ChildItem @()' -cursorColumn 1
```
Comment (daxian-dbw) [inline test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:822]: Why not use `Get-Chi<tab> @()` as is called out in the PR description?

```suggestion
        $res = TabExpansion2 -inputScript 'Get-Chi @()' -cursorColumn 7
```

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1 (hunk 2 above)
Comment (MartinGC94) [reply test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:822]: To make it less likely for the test to break if a new command gets added.

Comment (daxian-dbw) [reply test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:822]: I see. Then we are good.

**Review by daxian-dbw — APPROVED** [review]
LGTM

**Comment by daxian-dbw** [comment]
To maintainers: according to https://github.com/PowerShell/PowerShell/pull/18355#issuecomment-1288128716, this PR fixes a regression that was introduced in 7.3.0 preview 8. So, we may want to backport this fix to 7.3.0.

**Pull request merged**